include/deal.II/numerics/vector_tools.h

```diff
@@ -924,6 +933,29 @@ namespace VectorTools
                 VectorType &vec_result);
 
   /**
+   * Generic implementation for the project() function with FiniteElements that
+   * are supported by the MatrixFree class.
+   * Use this function if you use more than four components or the degree of
+   * your FiniteElement is higher than eight. You have to specify the first
+   * template arguments yourself.
+   * @p vec_result is expected to not have any ghost entries.
+   * @p project_to_boundary_first and @p enforce_zero_boundary are not yet
+   * implemented.
+   */
+  template <int components, int fe_degree, int dim, typename VectorType, int spacedim>
+  void project_generic (const Mapping<dim, spacedim>                              &mapping,
```

Comment:
but do we actually need that `project_generic`? why not do the switch internally in anonymous namespace in `templates.h`? I think we shall not confuse the user and provide `project()` which internally switches to matrix-free if it can and do something else otherwise.